x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java

```diff
@@ -0,0 +1,258 @@
+            persistentTask -> waitForJobStarted(persistentTask.getId(), params, request, listener),
+            e -> {
+                if (ExceptionsHelper.unwrapCause(e) instanceof ResourceAlreadyExistsException) {
+                    e = ExceptionsHelper.conflictStatusException(
+                        "Cannot upgrade job [{}] snapshot [{}] because upgrade is already in progress",
+                        e,
+                        request.getJobId(),
+                        request.getSnapshotId());
+                }
+                listener.onFailure(e);
+            });
+
+        // Start job task
+        ActionListener<Long> memoryRequirementRefreshListener = ActionListener.wrap(
+            mem -> {
+                logger.info("[{}] [{}] sending start upgrade request", params.getJobId(), params.getSnapshotId());
+                persistentTasksService.sendStartRequest(
+                    MlTasks.snapshotUpgradeTaskId(params.getJobId(), params.getSnapshotId()),
+                    MlTasks.JOB_SNAPSHOT_UPGRADE_TASK_NAME,
+                    params,
+                    waitForJobToStart);
+            },
+            listener::onFailure
+        );
+
+        // Check that model snapshot exists and should actually be upgraded
+        // Then refresh the memory
+        ActionListener<Result<ModelSnapshot>> getSnapshotHandler = ActionListener.wrap(
+            response -> {
+                if (response == null) {
+                    listener.onFailure(
+                        new ResourceNotFoundException(
+                            Messages.getMessage(Messages.REST_NO_SUCH_MODEL_SNAPSHOT, request.getSnapshotId(), request.getJobId())));
+                    return;
+                }
+                if (response.result.getMinVersion().major >= UPGRADE_FROM_MAJOR) {
+                    listener.onFailure(ExceptionsHelper.conflictStatusException(
+                        "Cannot upgrade job [{}] snapshot [{}] as it is already compatible with current major version {}",
+                        request.getJobId(),
+                        request.getSnapshotId(),
+                        UPGRADE_FROM_MAJOR));
+                    return;
+                }
+                memoryTracker.refreshAnomalyDetectorJobMemoryAndAllOthers(params.getJobId(), memoryRequirementRefreshListener);
+            },
+            listener::onFailure
+        );
+
+        ActionListener<Job> getJobHandler = ActionListener.wrap(
+            job -> {
+                if (request.getSnapshotId().equals(job.getModelSnapshotId())) {
+                    listener.onFailure(ExceptionsHelper.conflictStatusException(
+                        "Cannot upgrade snapshot [{}] for job [{}] as it is the current primary job snapshot",
+                        request.getSnapshotId(),
+                        request.getJobId()
+                    ));
```

Comment:
It doesn't make sense to upgrade the currently used snapshot of the job. If a job has been closed for an extended period of time (two major versions to be exact), it is probably a good idea to open it anyways for other reasons (job config updates, etc.)

Reply:
This means extra complication for the Kibana upgrade assistant though.  For every model snapshot that exists that is too old it will now have to recommend one of two possible courses of action, depending on whether the snapshot is the active one or not.  Opening and closing a job normally without sending it any data doesn't rewrite the snapshot, so the user would also have to feed some data to actually change the active model snapshot of the job.  So I think this check should be altered to only ban upgrading the active snapshot if the job is open.

Reply:
> So I think this check should be altered to only ban upgrading the active snapshot if the job is open.

100%, this seems good